crates/wasi-nn/src/backend/pytorch.rs

```diff
@@ -0,0 +1,174 @@
+        GraphEncoding::Pytorch
+    }
+
+    fn load(&mut self, builders: &[&[u8]], target: ExecutionTarget) -> Result<Graph, BackendError> {
+        if builders.len() != 1 {
+            return Err(BackendError::InvalidNumberOfBuilders(1, builders.len()).into());
+        }
+        // Load the torchscript saved module.
+        let mut saved_module = builders[0];
+
+        // Load the saved model on the device.
+        let mut compiled_module = CModule::load_data_on_device(
+            &mut saved_module,
+            map_execution_target_to_string(target),
+        )?;
+
+        // Set the model to be used for inference (eval), default mode is training.
+        compiled_module.f_set_eval()?;
+
+        let graph = PytorchGraph(Arc::new(Mutex::new(compiled_module)));
+        let box_: Box<dyn BackendGraph> = Box::new(graph);
+        Ok(box_.into())
+    }
+
+    fn as_dir_loadable<'a>(&'a mut self) -> Option<&'a mut dyn BackendFromDir> {
+        Some(self)
+    }
+}
+
+impl BackendFromDir for PytorchBackend {
+    fn load_from_dir(
+        &mut self,
+        path: &Path,
+        target: ExecutionTarget,
+    ) -> Result<Graph, BackendError> {
+        // Load the model from the file path
+        let compiled_module =
+            CModule::load_on_device(path, map_execution_target_to_string(target)).unwrap();
+        let graph = PytorchGraph(Arc::new(Mutex::new(compiled_module)));
+        let box_: Box<dyn BackendGraph> = Box::new(graph);
+        Ok(box_.into())
+    }
+}
+
+struct PytorchGraph(Arc<Mutex<tch::CModule>>);
+
+unsafe impl Send for PytorchGraph {}
+unsafe impl Sync for PytorchGraph {}
+
+impl BackendGraph for PytorchGraph {
+    fn init_execution_context(&self) -> Result<ExecutionContext, BackendError> {
+        let tensor = TchTensor::new();
+        let box_: Box<dyn BackendExecutionContext> =
+            Box::new(PytorchExecutionContext(self.0.clone(), tensor));
+        Ok(box_.into())
+    }
+}
+
+unsafe impl Sync for PytorchExecutionContext {}
+struct PytorchExecutionContext(Arc<Mutex<tch::CModule>>, tch::Tensor);
+
+impl BackendExecutionContext for PytorchExecutionContext {
+    fn set_input(&mut self, _index: Id, input_tensor: &Tensor) -> Result<(), BackendError> {
+        // Input index is not used in pytorch models. The forward method to a model passes the tensor/data to
+        // the appropriate layer of the model.
+        let kind = input_tensor.ty.try_into()?;
+        let dimensions = input_tensor
+            .dimensions
+            .iter()
+            .map(|&dim| dim as i64)
+            .collect::<Vec<_>>();
+        self.1 = TchTensor::from_data_size(&input_tensor.data, &dimensions, kind);
+        Ok(())
+    }
+
+    fn compute(&mut self) -> Result<(), BackendError> {
+        // Use forward method on the compiled module/model after locking the mutex, and pass the input tensor to it
+        self.1 = self.0.lock().unwrap().forward_ts(&[&self.1]).unwrap();
```

Comment:
Storing the output tensor in the same location as the input tensor means that `set_input` followed immediately by `get_output` would return the input tensor... probably not what you want here. It looks like `forward_ts` only returns a single tensor so perhaps just create an `output` field for that and another `input: Vec<Tensor>` for the inputs.

Reply:
Good point. I added `inputs` as vector of tensors. Also changed structs in general to use named fields to make code more readable.